scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -451,6 +462,32 @@ object OptimizationRules {
     }
   }
 
+  /**
+   * If we assume that Orderings are coherent, which we do generally in
+   * scalding in joins for instance, we can compose two reduce steps
+   */
+  object ComposeReduceSteps extends Rule[TypedPipe] {
+    def apply[A](on: Dag[TypedPipe]) = {
+      case ReduceStepPipe(rs2) =>
+        rs2.mapped match {
+          case ReduceStepPipe(rs1) =>
+            ReduceStep.maybeCompose(rs1, rs2).map(ReduceStepPipe(_))
+          case WithDescriptionTypedPipe(ReduceStepPipe(rs1), descs) =>
+            ReduceStep.maybeCompose(rs1, rs2).map { rs3 =>
+              WithDescriptionTypedPipe(ReduceStepPipe(rs3), descs)
+            }
+          case CoGroupedPipe(cg1) =>
+            CoGrouped.maybeCompose(cg1, rs2).map(CoGroupedPipe(_))
+          case WithDescriptionTypedPipe(CoGroupedPipe(cg1), descs) =>
```

Comment:
Is there a guarantee that there aren't nested descriptions here?

Reply:
There is not, but there is another optimization rule that collapses those, so this rule should run after that rule.